qiskit/quantum_info/synthesis/two_qubit_decompose.py

```diff
@@ -926,9 +949,227 @@ def __call__(self, target, basis_fidelity=None, *, _num_basis_uses=None) -> Quan
+
+        Returns:
+            QuantumCircuit: pulse optimal quantum circuit.
+            None: Probably nbasis=1 and original circuit is fine.
+
+        Raises:
+            QiskitError: Decomposition for selected basis not implemented.
+        """
+        circuit = None
+        if self.pulse_optimize and best_nbasis not in {0, 1}:
+            # already pulse optimal
+            pass
+        elif self.pulse_optimize and best_nbasis > 3:
+            raise QiskitError(
+                f"Unexpected number of entangling gates ({best_nbasis}) in decomposition."
+            )
+        if self._decomposer1q.basis in {"ZSX", "ZSXX"}:
+            if isinstance(self.gate, CXGate):
+                if best_nbasis == 3:
+                    circuit = self._get_sx_vz_3cx_efficient_euler(decomposition, target_decomposed)
+                elif best_nbasis == 2:
+                    circuit = self._get_sx_vz_2cx_efficient_euler(decomposition, target_decomposed)
+            else:
+                raise QiskitError("pulse_optimizer currently only works with CNOT entangling gate")
+        else:
+            raise QiskitError(
+                '"pulse_optimize" currently only works with ZSX basis '
+                f"({self._decomposer1q.basis} used)"
+            )
+        return circuit
+
+    def _get_sx_vz_2cx_efficient_euler(self, decomposition, target_decomposed):
+        """
+        Decomposition of SU(4) gate for device with SX, virtual RZ, and CNOT gates assuming
+        two CNOT gates are needed.
+
+        This first decomposes each unitary from the KAK decomposition into ZXZ on the source
+        qubit of the CNOTs and XZX on the targets in order to commute operators to beginning and
+        end of decomposition. The beginning and ending single qubit gates are then
+        collapsed and re-decomposed with the single qubit decomposer. This last step could be avoided
+        if performance is a concern.
+        """
+        best_nbasis = 2  # by assumption
+        num_1q_uni = len(decomposition)
+        # list of euler angle decompositions on qubits 0 and 1
+        euler_q0 = np.empty((int(num_1q_uni / 2), 3), dtype=float)
+        euler_q1 = np.empty((int(num_1q_uni / 2), 3), dtype=float)
+        global_phase = 0.0
+
+        # decompose source unitaries to zxz
+        zxz_decomposer = OneQubitEulerDecomposer("ZXZ")
+        for iqubit, idecomp in enumerate(range(0, num_1q_uni, 2)):
+            euler_angles = zxz_decomposer.angles_and_phase(decomposition[idecomp])
+            euler_q0[iqubit, [1, 2, 0]] = euler_angles[:3]
+            global_phase += euler_angles[3]
+        # decompose target unitaries to xzx
+        xzx_decomposer = OneQubitEulerDecomposer("XZX")
+        for iqubit, idecomp in enumerate(range(1, num_1q_uni, 2)):
+            euler_angles = xzx_decomposer.angles_and_phase(decomposition[idecomp])
+            euler_q1[iqubit, [1, 2, 0]] = euler_angles[:3]
+            global_phase += euler_angles[3]
```

Comment:
The function-local variable `global_phase` never seems to get used after this. Do you mean to incorporate it into `qc.global_phase` somehow?

Reply:
Yes. I guess it never caused in error though because the decomposition always seems to keep it near zero.